lib/src/main/java/space/taran/arklib/domain/index/RootIndex.kt

```diff
@@ -0,0 +1,135 @@
+ *    from all root folders.
+ * @param root A path to the root folder.
+ *
+ * See also [IndexAggregation] and [IndexProjection].
+ */
+@OptIn(ExperimentalPathApi::class)
+class RootIndex(val root: Path): ResourceIndex {
+
+    private val mutex = Mutex()
+    private val mutUpdates = MutableSharedFlow<ResourceUpdates>()
+
+    private val resourceById: MutableMap<ResourceId, Resource> = mutableMapOf()
+    private val pathById: MutableMap<ResourceId, Path> = mutableMapOf()
+
+    private fun wrap(update: RawUpdates): ResourceUpdates {
+        val deleted = update.deleted.associateWith { id ->
+            val path = pathById[id]!!
+            val resource = resourceById[id]!!
+            LostResource(path, resource)
+        }
+
+        val added = update.added
+            //we can't present empty resources
+            .mapNotNull { (id, path) ->
+                val resource: Resource = Resource.compute(id, path)
+                    .onFailure { error ->
+                        Log.e(
+                            LogTags.RESOURCES_INDEX,
+                            "Couldn't compute resource by path $path: $error"
+                        )
+
+                        return@mapNotNull null
+                    }.getOrThrow()
+
+                id to NewResource(path, resource)
+            }.toMap()
+
+        return ResourceUpdates(deleted, added)
+    }
+
+    init {
+        if (!BindingIndex.load(root)) {
+            Log.e(
+                RESOURCES_INDEX,
+                "Couldn't provide index from $root"
+            )
+            throw NotImplementedError()
+        }
+
+        //id2path should be used in order to filter-out duplicates
+        //path2id could contain several paths for the same id
+        BindingIndex.id2path(root)
+            .forEach { (id, path) ->
+                Resource.compute(id, path)
+                    .onFailure { error ->
+                        Log.e(
+                            RESOURCES_INDEX,
+                            "Couldn't compute resource by path $path: $error"
+                        )
+                    }
+                    .onSuccess { resource ->
+                        pathById[id] = path
+                        resourceById[id] = resource
+                    }
+            }
+    }
+
+    override val updates = mutUpdates.asSharedFlow()
+
+    override suspend fun updateAll(): Unit =
+        withContextAndLock(Dispatchers.IO, mutex) {
+            val raw: RawUpdates = BindingIndex.update(root)
+            BindingIndex.store(root)
+
+            val updates: ResourceUpdates = wrap(raw)
+
+            updates.deleted.forEach { (id, _) ->
+                resourceById.remove(id)
+                pathById.remove(id)
+            }
+
+            updates.added.forEach { (id, added) ->
+                resourceById[id] = added.resource
+                pathById[id] = added.path
+            }
+
+            mutUpdates.emit(updates)
```

Comment:
It looks fine, but if we'll have any problems we can create mutable flow on interface level (so `IndexAggregation` and `IndexProjection` would create their own mutable flows) and push updates into this flow **after** filtering/joining.